### DISCUSSION TITLE
How does Valtio causes re-renders without using useState ?

### DISCUSSION THREAD

**Discussion by afkcodes** [pmndrs/valtio#876] [General]
Hello @dai-shi  I have been using valtio for a while, and just like lot of people do i did not look into how valtio actually does what it does, i was thinking valtio might be using setState internally to trigger re-renders. but when i had a look into the react implementation i could not find it. I'm curious about how does it triggers re-renders in react, without useState or useReducer. 

Please correct me if I'm wrong above, i want to learn things. Thank you :)

**Comment by dai-shi** [comment]
It uses [useSyncExternalStore](https://react.dev/reference/react/useSyncExternalStore) provided in React 18, and `use-sync-external-store` has a shim for React 17.

See also: https://blog.axlight.com/posts/how-valtio-proxy-state-works-react-part/

**Reply by afkcodes** [reply]
Thank you for pointing it out, you have always helped in understanding and educating people. :)
